src/chainstate/burn/operations/delegate_stx.rs

```diff
@@ -0,0 +1,574 @@
+        let until_burn_height = {
+            if data[16] == 1 {
+                if data.len() < 25 {
+                    // too short to have required data
+                    warn!("DELEGATE_STX payload is malformed ({} bytes)", data.len());
+                    return None;
+                }
+                let burn_height = parse_u64_from_be(&data[17..25]).unwrap();
+                Some(burn_height)
+            } else if data[16] == 0 {
+                None
+            } else {
+                warn!("DELEGATE_STX payload is malformed (invalid byte value for option flag)");
+                return None;
+            }
+        };
+
+        Some(ParsedData {
+            delegated_ustx,
+            until_burn_height,
+        })
+    }
+
+    pub fn get_sender_txid(tx: &BurnchainTransaction) -> Result<&Txid, op_error> {
+        match tx.get_input_tx_ref(0) {
+            Some((ref txid, vout)) => {
+                if *vout != 1 {
+                    warn!("Invalid tx: DelegateStxOp must spend the second output of the PreStxOp");
+                    Err(op_error::InvalidInput)
+                } else {
+                    Ok(txid)
+                }
+            }
+            None => {
+                warn!("Invalid tx: DelegateStxOp must have at least one input");
+                Err(op_error::InvalidInput)
+            }
+        }
+    }
+
+    pub fn parse_from_tx(
+        block_height: u64,
+        block_hash: &BurnchainHeaderHash,
+        tx: &BurnchainTransaction,
+        sender: &StacksAddress,
+    ) -> Result<DelegateStxOp, op_error> {
+        let outputs = tx.get_recipients();
+
+        if tx.num_signers() == 0 {
+            warn!(
+                "Invalid tx: inputs: {}, outputs: {}",
+                tx.num_signers(),
+                outputs.len()
+            );
+            return Err(op_error::InvalidInput);
+        }
+
+        if outputs.len() == 0 {
+            warn!(
+                "Invalid tx: inputs: {}, outputs: {}",
+                tx.num_signers(),
+                outputs.len()
+            );
+            return Err(op_error::InvalidInput);
+        }
+
+        if tx.opcode() != Opcodes::DelegateStx as u8 {
+            warn!("Invalid tx: invalid opcode {}", tx.opcode());
+            return Err(op_error::InvalidInput);
+        };
+
+        let data = DelegateStxOp::parse_data(&tx.data()).ok_or_else(|| {
+            warn!("Invalid tx data");
+            op_error::ParseError
+        })?;
+
+        let delegate_to = outputs[0]
+            .address
+            .clone()
+            .try_into_stacks_address()
+            .ok_or_else(|| {
+                warn!("Invalid tx: output 1 must be representable as a StacksAddress");
+                op_error::InvalidInput
+            })?;
+
+        // coerce a hash mode for this address if need be, since we'll need it when we feed this
+        // address into the .pox contract
+        let reward_addr = if outputs.len() >= 2 {
```

Comment:
Hmm, I think the structure of the outputs here is problematic.  If you do _not_ want to set a PoX address (which is the usual behavior), then the way this is written, the Bitcoin transaction cannot have a change output.  This means that if you want to create such a transaction, all the BTC you spend from the inputs must be spent by this transaction (which is almost never what you want to do).

I think we'll need to address this in one of two ways:
* Make it so that `outputs[0]` is the `delegate-to` address; `outputs[1]` is the change address, and `outputs[2]`, if present, is the PoX address.  OR:
* Add another field to the `OP_RETURN` which indicates which output is the PoX address, if any.  You'd encode it the same way as the `until-burn-ht` value -- `0x00` for "no PoX address" and "0x01 0xXX` for "use the Bitcoin output at 0xXX as the PoX address".

Up to you how you want to proceed.  I'm more in favor of the second one since conventionally the change address is the last output in a Bitcoin transaction (and lots of wallets are written that way).

Reply:
Ahh... I see. I will do it the second way.